### PR TITLE
Flow 0.98.0

### PR DESCRIPTION
**Pull request by ray-kraesig**
Fix most, although sadly not all, errors reported by Flow 0.98.0.

A few such errors (mostly unused-suppression-comment warnings) remain, to be done in tandem with the upgrade itself.

In support of #3548.

### CODE REVIEW

File: src/webview/js/js.js [hunk 1]
```diff
@@ -216,7 +216,7 @@ function midMessagePeer(top: number, bottom: number): ?Element {
   }
 
   // $FlowFixMe: doesn't know about Document#elementsFromPoint
-  const midElements: Array<Element> = document.elementsFromPoint(0, midY);
+  const midElements: Array<HTMLElement> = document.elementsFromPoint(0, midY);
```
Comment (gnprice) [inline src/webview/js/js.js]: Should remove the fixme too, right?

... Ah, got it, this is all for a *future* version of Flow. :slightly_smiling_face: The commit message sounded like it was describing something already broken:

    flow: fix return type of `elementsFromPoint`
    
    `document.elementsFromPoint` was added in Flow v0.98.0. Bring our
    declared return type into compliance.
    
    https://github.com/facebook/flow/commit/c34fd59cf346650fbf2eeaedc38af9adf1d17da8

One thing I did a lot in the last round of this, which can help make this kind of story clearer, is to put the upcoming version in the future tense. So e.g. "will appear in Flow v0.98", or "Flow v0.98 will bring us a libdef for `document.elementsFromPoint`." (That Flow version already exists, but it hasn't yet appeared in our codebase, and will when we update to it.)

File: src/utils/jsonable.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  | number
+  | boolean
+  | { +[string]: JSONable } /* [α] */
+  | $ReadOnlyArray<JSONable>;
+
+// [α]: This should really be an exact type, `{| +[string]: JSONable |}`.
+// Unfortunately, it can't yet be.
+//
+// Prior to Flow v0.111.0 (i.e., prior to React Native v0.62.0), exact object
+// types with indexer properties are unusably broken. The following trivial
```
Comment (gnprice) [inline src/utils/jsonable.js]: That's good news that there *is* now a version of Flow where indexer properties on exact object types work reasonably! Please mention that on the RN v0.62 upgrade issue, too :slightly_smiling_face:

**Comment by gnprice** [comment]
> I like the idea of a `JSONable` type!

(And the implementation, too. It just occurred to me that that was ambiguous. :slightly_smiling_face: )

**Comment by ray-kraesig** [comment]
(The recent push rebased onto the latest `master` and adjusted some commit messages. All changes were to metadata; the content is identical.)

File: src/api/apiErrors.js [hunk 3]
```diff
@@ -45,7 +45,7 @@ export const makeErrorFromApi = (httpStatus: number, data: mixed): Error => {
 
   // Server has responded, but the response is not a valid error-object.
   // (This should never happen, even on old versions of the Zulip server.)
-  logging.warn(`Bad response from server: ${JSON.stringify(data)}`);
+  logging.warn(`Bad response from server: ${JSON.stringify(data) ?? 'undefined'}`);
   return new Error('Server responded with invalid message');
```
[the remark this answers is not in the document]
Comment (ray-kraesig) [reply src/api/apiErrors.js]: I am in turn confused by your explication of your confusion!

The Sentry logs act as a witness to the NFCness of this change by already including many, *many* instances of the error message "Bad response from server: undefined".

File: src/boot/TranslationProvider.js [hunk 4]
```diff
@@ -23,10 +23,10 @@ export const TranslationContext = React.createContext(undefined);
  * for the legacy context API.  When that isn't the case, simply saying
  * `context: TranslationContext` may be more convenient.
  */
-export function withGetText<P: { _: GetText }, C: ComponentType<P>>(
+export function withGetText<P: {| _: GetText |}, C: ComponentType<P>>(
   WrappedComponent: C,
 ): ComponentType<$Diff<ElementConfig<C>, { _: GetText }>> {
-  return class extends React.Component<$Diff<ElementConfig<C>, { _: GetText }>> {
+  return class extends React.Component<$Diff<ElementConfig<C>, {| _: GetText |}>> {
```
[the remark this answers is not in the document]
Comment (ray-kraesig) [reply src/boot/TranslationProvider.js]: Yeah, that's clearly not right. Will adjust.

File: src/api/apiErrors.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply src/api/apiErrors.js]: > The Sentry logs act as a witness to the NFCness of this change by already including many, _many_ instances of the error message "Bad response from server: undefined".

Hahahaha, I see. I guess maybe the genre-aware prediction in the first place would have been that of course there are events for it in the wild, that there are always events in the wild.

Maybe what'd help then is just to unpack that sentence a bit: as witnessed by our Sentry logs, this already comes out as "undefined" when the value is undefined, so this is not a functional change.

File: src/boot/TranslationProvider.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ray-kraesig) [reply src/boot/TranslationProvider.js]: Removing this change entirely causes no additional errors, so I've done that. (If whatever caused it pops up again during the RN 0.60 transition, I'll fold it in with the backwards-incompatible changes.)

**Comment by gnprice** [comment]
Looks good! Merging.

Made a tiny change to one commit message: in "Mark a few types as exact" did s/few/few more/, because on first read (I guess second read, counting the other day) it sounded like saying this was the beginning of marking our object types as exact, whereas in fact I think these are among the few stragglers.

**Pull request merged**